### DISCUSSION TITLE
exception_handlers not triggered during unit tests

### DISCUSSION THREAD

**Discussion by lays147** [fastapi/fastapi#9019] [Questions] [answered]
Hi, I'm using exception_handlers to handle 500 and 503 HTTP codes. However when unit testing, they are ignored. If I start the server and test the route, the handler works properly.

```python
# main.py
@app.exception_handler(Exception)
def handle_500_error(request: Request, exception: Exception) -> JSONResponse:
    return JSONResponse(
        status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
        content=json.dumps({"detail":'Internal Server Error, contact support'}),
    )

@app.exception_handler(OperationalError)
def handle_503_error(request: Request, exception: Exception) -> JSONResponse:
    return JSONResponse(
        status_code=status.HTTP_503_SERVICE_UNAVAILABLE,
        content=json.dumps({"detail":'Service Unavailable, contact support'}),
# function

@router.get("/{partner_id}")
def get_partner_by_id(partner_id: str, db: Session = Depends(get_db)):
    """
    Retrieve a partner from the database by given id
    :param [partner_id]: id to be searched
    :returns: partner data
    """
    db_partner = search_partner_by_id(db=db, partner_id=partner_id)
    if not db_partner:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND, detail="Partner not found"
        )
    return Partner.from_orm(db_partner).dict(by_alias=True)

# test


def test_get_partner_by_id_503(client: TestClient, valid_body: dict) -> None:
    with patch(
        "partner_api.routers.partners.search_partner_by_id", side_effect=OperationalError
    ):
        query = f"/partners/{valid_body['id']}"
        response = client.get(query)
        assert (
            response.status_code == status.HTTP_503_SERVICE_UNAVAILABLE
        ), response.json()
```

## Environment

Python 3.8.6
fastapi==0.63.0

**Comment by Kludex** [comment]
AFAIK You cannot handle pure `Exception`. Needs to be an inherited exception class.

A non asked tip:
```python
def get_partner_by_id(partner_id: str, db: Session = Depends(get_db)):
    """
    Retrieve a partner from the database by given id
    \f
    :param [partner_id]: id to be searched
    :returns: partner data
    """
```
Reference: https://fastapi.tiangolo.com/advanced/path-operation-advanced-configuration/#advanced-description-from-docstring

If you add reproducible examples, it's easy for people to help you. :kissing:

**Comment by lays147** [comment]
@Kludex thanks for the tip.
Ok, but that does not explain why both handlers work when I have the server running and testing via curl or why the handler for ``OperationalError`` does not work during the unit test.
Also, the code sample is reproducible, as you can see on the test I'm mocking the database function to raise the exception, one just need to setup the fast API app.

**Comment by aksel** [comment]
I think this is related to #2799

**Comment by nadworny** [comment]
hey @lays147, unfortunately, your code snippet isn't complete as the requirements and the imports are missing. Could you please provide it so that others can help?

**Comment by gusmith** [comment, marked as answer]
@lays147 
This may be indeed the same issue as #2799 
In my case, the solution was to update one of the argument from the `TestClient`: `raise_server_exceptions` which is by default set to True.
You are not showing how you build the test client, but if you create it this way:
```
client = TestClient(app, raise_server_exceptions=False)
```
the server should not raise the exceptions when they occur, but instead return the http error you are expecting.
In the docs from [starlette](https://www.starlette.io/testclient/)
>By default the TestClient will raise any exceptions that occur in the application. Occasionally you might want to test the content of 500 error responses, rather than allowing client to raise the server exception. In this case you should use client = TestClient(app, raise_server_exceptions=False).
